### ISSUE TITLE
FMU output causality messed up

### ISSUE THREAD

**Issue by AnHeuermann** [OpenModelica/OpenModelica#9782]
### Description

When translating

```modelica
model TestModel "test model"
  Real T;
  input Real Tref1 "reference Temperature";
  parameter Real alpha1 = 10;
  parameter Real T0 = 0 "start value for T";
  output Real q "output variable";
initial equation
  T = T0;
equation
   der(T) = alpha1*Tref1; 
   q = alpha1*(Tref1-T);
end TestModel;
```

`T` is the state variable and `q` is computed from input `Tref1` and state variable `T`
```
der(T) := alpha1*Tref1;
q := alpha1*(Tref1-T);
```

When creating a Co-Simulation FMU the resulting sorted equations look completely differently:

```
der(T) := alpha1 * Tref1; 
q_der := (-alpha1) * der(T);
der(outputAlias_q) := q_der;
q := outputAlias_q;
T := Tref1 - outputAlias_q / alpha1;
```

### Steps to Reproduce

```modelica
model TestModel "test model"
  Real T;
  input Real Tref1 "reference Temperature";
  parameter Real alpha1 = 10;
  parameter Real T0 = 0 "start value for T";
  output Real q "output variable";
initial equation
  T = T0;
equation
   der(T) = alpha1*Tref1; 
   q = alpha1*(Tref1-T);
end TestModel;
```

```modelica
model SomeInput
  output Real Tref1;
equation
  Tref1 = time^2;
end SomeInput;
```

```modelica
model connectedModel
  SomeInput someInput;
  TestModel testModel;
equation
  testModel.Tref1 = someInput.Tref1;
end connectedModel;
```

Now generate Co-Simulation FMUs for `TestModel` and `SomeInput` and connect them in an SSP model:
![image](https://user-images.githubusercontent.com/38031952/204254411-5fbbfd1b-09f9-4130-8b17-1ca08af03828.png)

The simulation result for `testModel.q` is constant at `zero` while `testModel.T` is correct.

### Expected Behavior

`connectedModelFMU.testModel.q` should be identical to ``connectedModel.testModel.q`.

### Version and OS

 - OpenModelica Version: OpenModelica v1.20.0-dev-393-g911cc77848 (64-bit)
 - OS: Windows 10, 64 bit

### Discussed in https://github.com/OpenModelica/OpenModelica/discussions/9768

<div type='discussions-op-text'>

<sup>Originally posted by **ANaumann85** November 25, 2022</sup>
Are input variables in modelica assumed to be constant in time? See for example the following simple model:

```mo
model testModel "test model"
  extends Modelica.Icons.Example;

  Real T;
  input Real Tref1 "reference Temperature";
  parameter Real alpha1 = 10;
  parameter Real T0 = 0 "start value for T";
  output Real q "output variable";
initial equation
  T = T0;
equation
   der(T) = alpha1*Tref1; 
   q = alpha1*(Tref1-T);
end testModel;
```

where Tref1 is assumed to be a time dependent input. 

When I generate an FMU in cosimulation mode with that model, I get completely wrong solutions. The root cause is, that the differentiates the equation for q under the assumption, that Tref is constant in time.

Is that expected behavior? How can I circumvent this assumption and generate an FMU, where the compiler does not apply this assumption?
</div>

**Comment by casella** [comment]
@AnHeuermann, I don't get why the output equation should get differentiated at all, adding another state to the system, regardless of the fact that the input is constant or time-varying. Why should output equations be differentiated in the first place?

**Comment by AnHeuermann** [comment]
Looks like the issue comes from `introduceOutputRealDerivatives`. This was done by @arun3688 for directional derivatives for FMI I think. Basically the output equation

```
q = alpha1*(Tref1-T);
```

get's replaced by its differentiated version, so `q` becomes a state. Problem is, that we don't have the derivative of input variable `Tref1`.  Instead of failing we generate the wrong equation
```
der(q) = (-alpha1) * der(T)
```
where the derivative for `Tref1` is missing.

@arun3688 do you know if the FMI standard says anything about differentiating FMU inputs?
When creating a model where I force omc to differentiate the input it complains (but fails in a stupid way, see #9802).

**Comment by arun3688** [comment]
@AnHeuermann yes `introduceOutputRealDerivatives` was done to get the first order derivative of Real outputs and set `maxOutputDerivativeOrder = 1` for FMI CoSimulation. But this module has nothing to do with the inputs and according to my knowledge the FMI standard does not say anything about differentiation inputs, Does it works if you disable the `introduceOutputRealDerivatives` and export the fmu ?

**Comment by kabdelhak** [comment]
@arun3688 if the output of which you want the derivative depends on an input you have to differentiate the input. That's the problem here. Running the model without co-simulation (which also deactivates `introduceOutputRealDerivatives`) works fine.

**Comment by kabdelhak** [comment]
@arun3688 the thing is that you force the compiler into index reduction. you are demanding the derivative of the output so it has to come from somewhere.

```
Equations (2, 2)
========================================
1/1 (1): $y_der = der(y)   [binding |0|0|0|0|]
2/2 (1): y = sin(time)   [binding |0|0|0|0|]
```
with unknowns `der(y)` and `$y_der`. There is no unknown in the second equation so it will be differentiated. If the RHS of that side depends on an unknown it has to be differentiated.

**Comment by arun3688** [comment]
@kabdelhak how should we proceed with this, because we need the output derivatives to support interpolation of inputs.

**Comment by casella** [comment]
> @kabdelhak how should we proceed with this, because we need the output derivatives to support interpolation of inputs.

@arun3688, I read from FMI 2.0.3, Section 4.2.1, page 104:
> In order to enable the slave to interpolate the continuous real inputs between communication steps, the derivatives of the inputs with respect to time can be provided. Also, higher derivatives can be set to allow higher order interpolation. Whether a slave is able to interpolate and therefore needs this information is provided by the capability attribute `canInterpolateInputs`.
```
fmi2Status fmi2SetRealInputDerivatives(fmi2Component c,
    const fmi2ValueReference vr[], size_ _t nvr,
    const fmi2Integer order[],
    const fmi2Real value[]); 
```
As I understand, if the master wants the slave to compute derivatives for interpolation, it should also set the derivatives of inputs, at least for those systems that have direct input-output feedthrough. It will do that using this function, and then the derivative value can be used for `der(Tref1)`.

Do I miss something?

**Comment by ANaumann85** [comment]
That proposal raises another question to me. Maybe it is  just wording.

> > In order to enable the slave to interpolate the continuous real inputs between communication steps, the derivatives of the inputs with respect to time **can** be provided. Also, higher derivatives **can** be set to allow higher order interpolation. Whether a slave is able to interpolate and therefore needs this information is provided by the capability attribute `canInterpolateInputs`.
> As I understand, if the master wants the slave to compute derivatives for interpolation, it **should** also set the derivatives of inputs, at least for those systems that have direct input-output feedthrough. It will do that using this function, and then the derivative value can be used for `der(Tref1)`.
> 
> Do I miss something?

I added some emphasis. If the compiler differentiates the inputs, then the master **HAS TO** provide the derivatives. But where should the master get that information from? Especially, up to which derivative has to be provided?

Maybe another approach would be more feasible. If the compiler generates new variables for the derivatives, then the master could track their relation ship back. Then the omc should generate the corresponding code for the interpolation.

What do you think?

**Comment by casella** [comment]
FMI is so exquisitely complicated that it makes me cry 😅.

By reading Section 4.3.1 of FMI 2.0.3 I understand that the master algorithm can check the XML file for the `CoSimulation` element, and look for the `canInterpolateInputs` attribute. If this is true, then it can look at the `maxOutputDerivativeOrder` attribute to figure out what it the highest derivative that can be handled. This information has to be consistent with the actual capabilities of the generated FMU code.

> I added some emphasis. If the compiler differentiates the inputs, then the master **has to** provide the derivatives.

Not necessarily. It could be doing it in its spare time (that a [Monty Python joke](https://www.youtube.com/watch?v=ohDB5gbtaEQ&t=158s)) 😄 

Seriously, first the master checks the XML and understand if it can handle derivatives. If the answer is 'yes', then it _can_ call `fmi2SetRealInputDerivatives` and then `fmi2GetRealOutputDerivatives` to have them computed. But it can also _not do that_, and make do without input/output derivatives. If you don't call `fmi2GetRealOutputDerivatives`, the part of the FMU code depending on `der(Tref)` will never be executed, or anyway its results will never be enquired by the master, so it is ok if `der(Tref1)` is just initialized to zero and keeps that value. Who cares.

It's a choice of the master whether or not to use derivatives.

> But where should the master get that information from? Especially, up to which derivative has to be provided?

It is up to the master to decide if and how much it would like to interpolate. Then, of course, it needs to check if _all the supplied FMUs_ actually support that, and possibly to come up with the sad reality that the FMU can interpolate less, or maybe can't do that at all, and adapt to that reality. 

Do I miss something?

**Comment by ANaumann85** [comment]
Well, why are we talking about the derivatives of the **outputs**? The https://github.com/OpenModelica/OpenModelica/issues/9782#issuecomment-1330550361 shows, that the derivatives of the **inputs** are missing. 
I understood the https://github.com/OpenModelica/OpenModelica/issues/9782#issuecomment-1330848370 such, that the computation of the output derivatives only triggers the differentiation. 
At which point does the original model in https://github.com/OpenModelica/OpenModelica/issues/9782#issue-1466206134 ask for output derivatives? 

@arun3688 : can you please elaborate a bit in this? Maybe one can disable the computation of output derivatives as a work-around?

**Comment by casella** [comment]
> Well, why are we talking about the derivatives of the **outputs**? The [#9782 (comment)](https://github.com/OpenModelica/OpenModelica/issues/9782#issuecomment-1330550361) shows, that the derivatives of the **inputs** are missing.

@ANaumann85 I agree it is missing. If we want to support the computation of output derivatives for interpolation (which I understand is what @arun3688 is currently working on), then the input derivatives should be included by the backend in the differentiated output equation, and then their values set by the fmi2SetRealInputDerivatives FMI2 function.

> I understood the [#9782 (comment)](https://github.com/OpenModelica/OpenModelica/issues/9782#issuecomment-1330848370) such, that the computation of the output derivatives only triggers the differentiation. At which point does the original model in [#9782 (comment)](https://github.com/OpenModelica/OpenModelica/issues/9782#issue-1466206134) ask for output derivatives?

It's not the model. The point is, we want to generate FMUs that are able to compute the output derivatives, just in case the master solver asks for them. Given the output

```modelica
der(T) := alpha1 * Tref1; 
q_der := (-alpha1) * der(T);
der(outputAlias_q) := q_der;
q := outputAlias_q;
T := Tref1 - outputAlias_q / alpha1
```
I see three problems.

1. During differentiation `Tref1` is assumed as a constant. Since in general it isn't, we should add `Tref1_der` when differentiating `Tref1`, so that the master solver can set it by calling fmi2SetRealInputDerivatives 
2. The differentiation process eliminates the original algebraic output equations and introduces an extra state variable `outputAlias_q`. This is not a good idea because it can cause drift due to numerical integration errors. We should instead get something like this:
```modelica
der(T) := alpha1 * Tref1; 
q := alpha1*(Tref1-T);
q_der := alpha1 * (Tref1_der - der(T)); 
```
3. For some reason the co-simulation output computes `T`, which is instead a state variable, from the new state outputAlias_q. This makes little sense to me. Computing output derivatives shouldn't involve a state variable change, right?

I guess the same symbolic machinery that is used for index reduction and state variable change was used in order to get the output derivatives, but probably it would need some adaptations to work well in this context.

**Comment by arun3688** [comment]
@casella the support to compute output derivatives (first order derivatives)  for interpolation is already implemented, but this is not yet available  `then the input derivatives should be included by the backend in the differentiated output equation, and then their values set by the fmi2SetRealInputDerivatives FMI2 function.`, As you said the point is we want to generate fmu's which has the capability to support interpolation and our fmus does that, this feature has been tested by EDF,  but this feature is not tested intensively

**Comment by casella** [comment]
I understand adding the derivatives of inputs is essential, but not enough. As I wrote above, I see also some fishy dummy derivative stuff going on in the generated FMU-CS code, that IMO should be handled differently. Maybe we can discuss this briefly in a meeting. Unfortunately I won't be available on Monday, but we can do it later next week.

**Comment by arun3688** [comment]
@AnHeuermann @kabdelhak It would be good if we can come up with a small example and what type of new equations should be added if output depends on input
